pymc_experimental/tests/test_model_builder.py

```diff
@@ -0,0 +1,128 @@
+
+        a = pm.Normal("a", a_loc, sigma=a_scale)
+        b = pm.Normal("b", b_loc, sigma=b_scale)
+        obs_error = pm.HalfNormal("σ_model_fmc", obs_error)
+
+        y_model = pm.Normal("y_model", a + b * x, obs_error, observed=y_data)
+
+    def _data_setter(self, data: pd.DataFrame):
+        with self.model:
+            pm.set_data({"x": data["input"].values})
+            try:
+                pm.set_data({"y_data": data["output"].values})
+            except:
+                pm.set_data({"y_data": np.zeros(len(data))})
+
+    @classmethod
+    def create_sample_input(cls):
+        x = np.linspace(start=0, stop=1, num=100)
+        y = 5 * x + 3
+        data = pd.DataFrame({"input": x, "output": y})
+
+        model_config = {
+            "a_loc": 7,
+            "a_scale": 3,
+            "b_loc": 5,
+            "b_scale": 3,
+            "obs_error": 2,
+        }
+
+        sampler_config = {
+            "draws": 1_000,
+            "tune": 1_000,
+            "chains": 1,
+            "target_accept": 0.95,
+        }
+
+        return data, model_config, sampler_config
+
+
+def test_model_pickle():
+    data, model_config, sampler_config = test_ModelBuilder.create_sample_input()
+    model = test_ModelBuilder(model_config, sampler_config)
+    cloned = cloudpickle.loads(cloudpickle.dumps(model))
+    assert_equal(model.basic_RVs, cloned.basic_RVs)
+    assert_equal(model.id(), cloned.id())
+
+
+def test_fit():
+    with pm.Model() as model:
+        x = np.linspace(start=0, stop=1, num=100)
+        y = 5 * x + 3
+        x = pm.MutableData("x", x)
+        y_data = pm.MutableData("y_data", y)
+
+        a_loc = 7
+        a_scale = 3
+        b_loc = 5
+        b_scale = 3
+        obs_error = 2
+
+        a = pm.Normal("a", a_loc, sigma=a_scale)
+        b = pm.Normal("b", b_loc, sigma=b_scale)
+        obs_error = pm.HalfNormal("σ_model_fmc", obs_error)
+
+        y_model = pm.Normal("y_model", a + b * x, obs_error, observed=y_data)
+
+        idata = pm.sample(1000, tune=1000)
+    data, model_config, sampler_config = test_ModelBuilder.create_sample_input()
+    model_2 = test_ModelBuilder(model_config, sampler_config)
+
+    assert_equal(model_2.idata, idata)
+
+
+def test_predict():
+    with pm.Model() as model:
+        x = np.linspace(start=0, stop=1, num=100)
+        y = 5 * x + 3
+        x = pm.MutableData("x", x)
+        y_data = pm.MutableData("y_data", y)
+        a_loc = 7
+        a_scale = 3
+        b_loc = 5
+        b_scale = 3
+        obs_error = 2
+
+        a = pm.Normal("a", a_loc, sigma=a_scale)
+        b = pm.Normal("b", b_loc, sigma=b_scale)
+        obs_error = pm.HalfNormal("σ_model_fmc", obs_error)
+
+        y_model = pm.Normal("y_model", a + b * x, obs_error, observed=y_data)
+
+        idata = pm.sample(1000, tune=1000)
+        y_test = pm.sample_posterior_predictive(idata)
+
+        y_test.posterior_predictive.equlas(model.posterior_predictive)
```

Comment:
Needs an assert. And `equlas`?

Reply:
`assert` can only work with certain datatypes. `idata` itself is an `xarray.Dataset` which doesn't not work with `assert_equal` . This can be replaced using this. link for documentation of [equals](https://docs.xarray.dev/en/v0.12.3/generated/xarray.Dataset.equals.html)

Reply:
But it spells `equlas`, did you run the tests to see if they pass?

Reply:
For the equals assert here, you don't need to run 2000 iterations.

Please do `pm.sample(tune=10, draws=20, chains=3, cores=1, step=pm.Metropolis(), compute_convergence_checks=False)` here and above, so the test runtime doesn't increase too much.